src/backend/distributed/utils/metadata_cache.c

```diff
@@ -1119,30 +1205,101 @@ ErrorIfAvailableVersionMismatch(void)
  * major or minor version numbers, this function errors out. It ignores the schema
  * version.
  */
-static void
+void
 ErrorIfInstalledVersionMismatch(void)
 {
-	char *installedVersion = NULL;
-
 	if (!EnableVersionChecks)
 	{
 		return;
 	}
 
-	installedVersion = InstalledExtensionVersion();
-	if (!MajorVersionsCompatible(installedVersion, CITUS_EXTENSIONVERSION))
+	if (!InstalledVersionCompatible())
 	{
 		ereport(ERROR, (errmsg("loaded Citus library version differs from installed "
 							   "extension version"),
 						errdetail("Loaded library requires %s, but the installed "
 								  "extension version is %s.", CITUS_MAJORVERSION,
-								  installedVersion),
+								  InstalledExtensionVersion()),
 						errhint("Run ALTER EXTENSION citus UPDATE and try again.")));
 	}
 }
 
 
 /*
+ * AvailableVersionCompatible returns whether available extension version is compatible
+ * with loaded library version. It also saves the result, thus consecutive calls to
+ * AvailableVersionCompatible will not parse the version strings again.
+ */
+static bool
+AvailableVersionCompatible(void)
+{
+	char *availableVersion = NULL;
+	MemoryContext oldMemoryContext = NULL;
+
+	/* if we cached the result before, return it */
+	if (availableVersionCompatible != NULL)
+	{
+		return *availableVersionCompatible;
+	}
+
+	availableVersion = AvailableExtensionVersion();
+
+	/* we will cache the result of comparision to prevent string parsing */
+	if (CacheMemoryContext == NULL)
+	{
+		CreateCacheMemoryContext();
+	}
+
+	oldMemoryContext = MemoryContextSwitchTo(CacheMemoryContext);
+
+	availableVersionCompatible = palloc0(sizeof(bool));
+	*availableVersionCompatible = MajorVersionsCompatible(availableVersion,
```

Comment:
A pointer here is pretty awkward - why don't you instead just redo the check as long as availableVersionCompatible == false, and set it to true the first time it passes?